### PR TITLE
Add network configuration

### PR DESCRIPTION
**Pull request by bdraco**
## Proposed change

https://github.com/home-assistant/architecture/discussions/509
https://github.com/home-assistant/core/pull/50874

Screen Shot 2021-05-19 at 4 59 20 PM
Screen Shot 2021-05-19 at 4 59 26 PM

## Type of change

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (thank you!)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example configuration

```yaml

```

## Additional information

- This PR fixes or closes issue: fixes #
- This PR is related to issue or discussion:
- Link to documentation pull request:

## Checklist

- [ ] The code change is tested and works locally.
- [ ] There is no commented out code in this PR.
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

[docs-repository]: https://github.com/home-assistant/home-assistant.io

### CODE REVIEW

**Comment by KapJI** [comment]
Is it possible to show next to auto configure option what it is selected?

**Comment by bdraco** [comment]
> Is it possible to show next to auto configure option what interface it has selected?

Yes, done.  Screen shots updated

**Comment by KapJI** [comment]
Sorry, this is not what I meant. My question was about if this UI can show what interface is used by auto detection.

**Comment by thecode** [comment]
> > Is it possible to show next to auto configure option what interface it has selected?
> 
> Yes, done. Screen shots updated

Where?
If I understood @KapJI he was (at least I was) expecting to see something like this:

- [X] Auto Configure (eth0 selected)

File: src/components/ha-network.ts [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+  );
+}
+
+@customElement("ha-network")
+export class HaNetwork extends LitElement {
+  @property({ attribute: false }) public hass!: HomeAssistant;
+
+  @property({ attribute: false }) public networkConfig?: NetworkConfig;
+
+  @property({ attribute: false }) private _expanded?: boolean;
```
Comment (bramkragten) [inline src/components/ha-network.ts:34]:
```suggestion
  @state() private _expanded?: boolean;
```

File: src/components/ha-network.ts [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+
+  private _handleAutoConfigureCheckboxClick(ev: Event) {
+    const checkbox = ev.currentTarget as HaCheckbox;
+    if (this.networkConfig === undefined) {
+      return;
+    }
+
+    if (checkbox.checked) {
+      this._expanded = false;
+      this.networkConfig.configured_adapters = [];
```
Comment (bramkragten) [inline src/components/ha-network.ts:93]: You should not change the data you get passed, instead, fire an event with an altered copy of the data

**Comment by bdraco** [comment]
> > > Is it possible to show next to auto configure option what interface it has selected?
> > 
> > 
> > Yes, done. Screen shots updated
> 
> Where?
> If I understood @KapJI he was (at least I was) expecting to see something like this:
> 
> * [x]  Auto Configure (eth0 selected)

Adjusted again

**Comment by chemelli74** [comment]
Just one small note: is not only for multicast traffic.
If we go in the direction to ban opening socket on "0.0.0.0" then you are forced to specify an IP. That means all traffic is affected by the change.

Simone

**Comment by bdraco** [comment]
> Just one small note: is not only for multicast traffic.
> 
> If we go in the direction to ban opening socket on "0.0.0.0" then you are forced to specify an IP. That means all traffic is affected by the change.
> 
> 
> 
> Simone
> 
> 

I don't think we are going to ban opening a socket on 0.0.0.0.  This change is more about letting integrations make safer networking choices in line with what the user wants

File: src/components/ha-network.ts (hunk 2 above)
Comment (bdraco) [reply src/components/ha-network.ts]: done.

File: src/components/ha-network.ts [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+import "./ha-settings-row";
+import "./ha-icon";
+
+function format_addresses(
+  addresses: IPv6ConfiguredAddress[] | IPv4ConfiguredAddress[]
+) {
+  return addresses.map(
+    (address) => html`<span>${address.address}/${address.network_prefix}</span>`
+  );
+}
```
Comment (bramkragten) [inline src/components/ha-network.ts:24]:
```suggestion
const format_addresses = (
  addresses: IPv6ConfiguredAddress[] | IPv4ConfiguredAddress[]
): TemplateResult => addresses.map(
    (address) => html`<span>${address.address}/${address.network_prefix}</span>`
  )
```

File: src/components/ha-network.ts [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+
+function format_auto_detected_interfaces(adapters: Adapter[]) {
+  return adapters.map((adapter) =>
+    adapter.auto
+      ? html`${adapter.name} (${format_addresses(adapter.ipv4)}
+        ${format_addresses(adapter.ipv6)} )`
+      : ""
+  );
+}
```
Comment (bramkragten) [inline src/components/ha-network.ts:33]:
```suggestion
const format_auto_detected_interfaces = (adapters: Adapter[]): TemplateResult | string => adapters.map((adapter) =>
    adapter.auto
      ? html`${adapter.name} (${format_addresses(adapter.ipv4)}
        ${format_addresses(adapter.ipv6)} )`
      : ""
  )
```

**Review by bramkragten — APPROVED** [review]

**Comment by bdraco** [comment]
Thanks 👍

File: src/components/ha-network.ts [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+import { haStyle } from "../resources/styles";
+import { HomeAssistant } from "../types";
+import "./ha-checkbox";
+import type { HaCheckbox } from "./ha-checkbox";
+import "./ha-settings-row";
+import "./ha-icon";
+
+const format_addresses = (
+  addresses: IPv6ConfiguredAddress[] | IPv4ConfiguredAddress[]
+): TemplateResult => addresses.map(
```
Comment (bramkragten) [inline src/components/ha-network.ts:20]:
```suggestion
): TemplateResult[] => addresses.map(
```

File: src/components/ha-network.ts [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+    (address) => html`<span>${address.address}/${address.network_prefix}</span>`
+  )
+
+const format_auto_detected_interfaces = (adapters: Adapter[]): TemplateResult | string => adapters.map((adapter) =>
```
Comment (bramkragten) [inline src/components/ha-network.ts:24]:
```suggestion
const format_auto_detected_interfaces = (adapters: Adapter[]): Array<TemplateResult | string> => adapters.map((adapter) =>
```

Comment (bdraco) [reply src/components/ha-network.ts]: Oops.  Pushed a different change before I saw this.  Reverted to this

File: src/panels/config/core/ha-config-section-core.js [hunk 7]
```diff
@@ -30,6 +31,7 @@ class HaConfigSectionCore extends LocalizeMixin(PolymerElement) {
         <ha-config-name-form hass="[[hass]]"></ha-config-name-form>
         <ha-config-core-form hass="[[hass]]"></ha-config-core-form>
         <ha-config-url-form hass="[[hass]]"></ha-config-url-form>
+        <ha-config-network hass="[[hass]]"></ha-config-network>
```
Comment (balloob) [inline src/panels/config/core/ha-config-section-core.js:34]: I think that this one (and the one above it) should both be hidden unless a user has advanced mode enabled.

Comment (bdraco) [reply src/panels/config/core/ha-config-section-core.js:34]: Url is already hidden unless they are in advanced mode.

I'll put the same change in network

Comment (bdraco) [reply src/panels/config/core/ha-config-section-core.js:34]: 97a8aa9f0

**Review by balloob — APPROVED** [review]

**Pull request merged**